src/main/java/donna/Donna.java

```diff
@@ -0,0 +1,180 @@
+package donna;
+
+import donna.task.Deadline;
+import donna.task.Event;
+import donna.task.Task;
+import donna.task.ToDo;
+
+import java.util.List;
+import java.util.Scanner;
+
+/**
+ * Represents the main class for the Donna chatbot.
+ */
+public class Donna {
+    private final Ui ui;
+    private TaskList tasks;
+    private final Parser parser;
+    private static Storage storage;
+
+    private static final String DIRECTORY_PATH = "data";
+    private static final String FILE_PATH = DIRECTORY_PATH + "/donna-tasks.txt";
+
+    /**
+     * Constructs a Donna instance, initializing the user interface, storage,
+     * parser, and loading previously saved tasks.
+     */
+    public Donna() {
+        ui = new Ui();
+        storage = new Storage(DIRECTORY_PATH, FILE_PATH);
+        parser = new Parser();
+        try {
+            tasks = new TaskList(storage.loadTasks());
+        } catch (DonnaException e) {
+            ui.printErrorMessage(e.getMessage());
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Runs the main loop, processing user input and executing commands.
+     * Handles exceptions and prints error messages as needed.
+     */
+    private void run() {
+        boolean dataWasLoaded = !tasks.isEmpty();
+        ui.printGreeting(dataWasLoaded);
+
+        Scanner sc = new Scanner(System.in);
+        while (true) {
+            String input = sc.nextLine();
+            try {
+                ParsedCommand result = parser.parse(input);
+                String commandType = result.getCommandType();
+
+                switch (commandType) {
+                case "exit":
+                    ui.printGoodbyeMessage();
+                    storage.saveTasks(tasks);
+                    sc.close();
+                    return;
+                case "list":
+                    ui.printTaskList(tasks);
+                    break;
+                case "mark":
+                    handleMark(result.getArgument1());
```

Comment:
I really like that you separate your command into arguments and command name in the ParsedCommand - it's very neat.

Reply:
Thank you!